src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,272 @@
+export type StreamName<T extends StreamType = StreamType> = `${T}:${string}`;
+
+export type StreamNameParts<T extends StreamType = StreamType> = {
+  streamType: T;
+  entityId: string;
+};
+
+export type StreamToProject<EventType extends Event> = {
+  streamName: StreamName;
+  streamType: StreamType;
+  entityId: string;
+  streamVersion: number;
+  events: ReadEvent<EventType, ReadEventMetadata>[];
+};
+
+export type Projection<EventType extends Event> = (
+  stream: StreamToProject<EventType>,
+) => void | Promise<void>;
+
+export interface EventStream<EventType extends Event> {
+  streamName: string;
+  events: Array<ReadEvent<EventType, ReadEventMetadata>>;
+  createdAt: Date;
+  updatedAt: Date;
+}
+export type EventStreamEvent<EventType extends Event> =
+  EventStream<EventType>['events'][number];
+
+export interface MongoDBConnectionOptions {
+  connectionString: string;
+  database: string;
+  collection?: string;
+}
+
+class EventStoreClass implements EventStore<number> {
+  private readonly collection: Collection<EventStream<Event>>;
+
+  constructor(collection: typeof this.collection) {
+    this.collection = collection;
+  }
+
+  async readStream<EventType extends Event>(
+    streamName: StreamName,
+    options?: ReadStreamOptions<number>,
+  ): Promise<Exclude<ReadStreamResult<EventType, number>, null>> {
+    const expectedStreamVersion = options?.expectedStreamVersion;
+
+    const stream = await this.collection.findOne<
+      WithId<EventStream<EventType>>
+    >({
+      streamName: { $eq: streamName },
+    });
+
+    if (!stream) {
+      return {
+        events: [],
+        currentStreamVersion: MongoDBEventStoreDefaultStreamVersion,
+        streamExists: false,
+      };
+    }
+
+    assertExpectedVersionMatchesCurrent(
+      stream.events.length,
+      expectedStreamVersion,
+      MongoDBEventStoreDefaultStreamVersion,
+    );
+
+    return {
+      events: stream.events.slice(0, maxEventIndex(expectedStreamVersion)),
+      currentStreamVersion: stream.events.length,
+      streamExists: true,
+    };
+  }
+
+  async aggregateStream<State, EventType extends Event>(
+    streamName: StreamName,
+    options: AggregateStreamOptions<State, EventType, number>,
+  ): Promise<AggregateStreamResult<State, number>> {
+    const stream = await this.readStream<EventType>(streamName, options?.read);
+    const state = stream.events.reduce(options.evolve, options.initialState());
+    return {
+      state,
+      currentStreamVersion: stream.currentStreamVersion,
+      streamExists: stream.streamExists,
+    };
+  }
+
+  async appendToStream<EventType extends Event>(
+    streamName: StreamName,
+    events: EventType[],
+    options?: AppendToStreamOptions<number> & {
+      /**
+       * These will be ran after a the events have been successfully appended to
+       * the stream. `appendToStream` will return after every projection is completed.
+       */
+      projections?: Array<Projection<EventType>>;
```

Comment:
**QUESTION:** @alex-laycalvert, could you expand on why you have the projections on the append level instead of the event store options (so as we have in PostgreSQL)?

Reply:
Let me look at the PostgreSQL example a bit more, there wasn't any specific reason I did this.